### PR TITLE
Changed column type of timestamp in LocalChangeEntity

### LINKED ISSUE #2083: Store timeStamp as Instant instead of human readable string format in LocalChangeEntity.
Planning to do it in a separate PR.

_Originally posted by @aditya-07 in https://github.com/google/android-fhir/pull/2030#discussion_r1268159725_

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2083

**Description**
 1. LocalChangeEntity.timestamp is change from human readable string to Instant like in other tables in the engine's db.
 2. Added Migration and tests. 

**Alternative(s) considered**
NO

**Type**
Choose one:  Code health 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/LocalChange.kt [hunk 1]
```diff
@@ -28,7 +29,7 @@ data class LocalChange(
   /** This is the id of the version of the resource that this local change is based of */
   val versionId: String? = null,
   /** last updated timestamp on server when this local changes are sync with server */
-  val timestamp: String = "",
+  val timestamp: Instant,
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/LocalChange.kt:32]: can you update the Kdoc for the `timestamp` variable?

**Review by omarismail94 — APPROVED** [review]
Thanks for this! Just one comment

File: engine/src/main/java/com/google/android/fhir/LocalChange.kt [hunk 2]
```diff
@@ -28,7 +28,7 @@ data class LocalChange(
   val resourceId: String,
   /** This is the id of the version of the resource that this local change is based of */
   val versionId: String? = null,
-  /** last updated timestamp on server when this local changes are sync with server */
+  /** This signifies the instant a CUD operation was performed on the resource. */
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/LocalChange.kt:31]: What do you think of this:
The time instant the app user performed a CUD operation on the resource.

File: engine/src/main/java/com/google/android/fhir/LocalChange.kt [hunk 3]
```diff
@@ -27,8 +28,8 @@ data class LocalChange(
   val resourceId: String,
   /** This is the id of the version of the resource that this local change is based of */
   val versionId: String? = null,
-  /** last updated timestamp on server when this local changes are sync with server */
-  val timestamp: String = "",
+  /** This signifies the instant a CUD operation was performed on the resource. */
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/LocalChange.kt:31]:
```suggestion
  /** The time instant the app user performed a CUD operation on the resource. */
```

**Review by omarismail94 — APPROVED** [review]

File: engine/src/main/java/com/google/android/fhir/db/impl/ResourceDatabase.kt [hunk 4]
```diff
@@ -108,3 +108,21 @@ val MIGRATION_4_5 =
...
+
+/** Changes column type of [LocalChangeEntity.timestamp] from [String] to [java.time.Instant]. */
+val MIGRATION_5_6 =
+  object : Migration(/* startVersion = */ 5, /* endVersion = */ 6) {
+    override fun migrate(database: SupportSQLiteDatabase) {
+      database.execSQL(
+        "CREATE TABLE IF NOT EXISTS `_new_LocalChangeEntity` (`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, `resourceType` TEXT NOT NULL, `resourceId` TEXT NOT NULL, `timestamp` INTEGER NOT NULL, `type` INTEGER NOT NULL, `payload` TEXT NOT NULL, `versionId` TEXT)"
+      )
+      database.execSQL(
+        "INSERT INTO `_new_LocalChangeEntity` (`id`,`resourceType`,`resourceId`,`timestamp`,`type`,`payload`,`versionId`) SELECT `id`,`resourceType`,`resourceId`, strftime('%s', `timestamp`) ||  substr(strftime('%f', `timestamp`), 4),`type`,`payload`,`versionId` FROM `LocalChangeEntity`"
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/ResourceDatabase.kt:120]: I'm wondering about if the string is (for some reason unclear to us) not convertible to a timestamp - how can we mitigate that? can we provide some level of error tolerance or at least recoverability for the client.

**Review by jingtang10 — COMMENTED** [review]
thanks aditya - great change!

**Pull request merged**